### ISSUE TITLE
Super expression must either be null or a function, not undefined

### ISSUE THREAD

**Issue by Vagabondtq** [sweetalert2/sweetalert2#1127]
Hi,
i ran sweetalert2 with npm and Laravel mix.

in my scrpit:
`window.swal = require('sweetalert2');`

and i use mixin:

```
swal.mixin({
  input: 'text',
  confirmButtonText: 'Next &rarr;',
  showCancelButton: true,
  progressSteps: ['1', '2', '3']
}).queue([
  {
    title: 'Question 1',
    text: 'Chaining swal2 modals is easy'
  },
  'Question 2',
  'Question 3'
]).then((result) => {
  if (result.value) {
    swal({
      title: 'All done!',
      html:
        'Your answers: <pre>' +
          JSON.stringify(result.value) +
        '</pre>',
      confirmButtonText: 'Lovely!'
    })
  }
})
```

build project with command: `npm run production`
I got error: 

> Super expression must either be null or a function, not undefined

Please help me.

**Comment by paolog22** [comment]
same. this is already reported but got nothing.

**Comment by zenflow** [comment]
@paolog22 It is already reported? Which issue is it? This is the first I recall hearing of it.

**Comment by limonte** [comment]
In our [Gitter](https://gitter.im/sweetalert2/Lobby?at=5b1a348eddd61d08e53c4401). There's a repo for reproduction: https://github.com/paolog22/sweetalert-nuxtjs-demo

I didn't have time to investigate, but my guess is that `nuxtjs` internally using `uglifyjs` which was causing the same issue for us until it was replaced by  `babel-minify` in #1062

**Comment by BrOrlandi** [comment]
I have the same error here after I build it for production.

**Comment by paolog22** [comment]
@limonte nuxt uses 

> "uglifyjs-webpack-plugin": "^1.2.5"

**Comment by limonte** [comment]
# Temporary workaround

Depending on your framework, find out the way to set the `unused` option of used `whatnot-uglify` plugin to `false`, e.g.

```js
$.uglify({
  compress: {
    unused: false
  }
})
```

### Technical details 

Apparently, `uglify-js/es` isn't supported anymore and will be replaced by `terser` in the near future: 

- https://github.com/webpack-contrib/uglifyjs-webpack-plugin/issues/262
- https://github.com/webpack-contrib/uglifyjs-webpack-plugin/pull/296

To clarify, the issue is happening when two of those things are collapsing together: 
- `unused` configuration option of UglifyJS2: https://github.com/mishoo/UglifyJS2#compress-options
- SweetAlert2's mixins

If the `unused` option is disabled, everything works fine, proof: https://github.com/sweetalert2/sweetalert2/tree/uglify-js-unused-false

@zenflow will you have time to investigate further the `uglify-js-unused-false` branch and figure out which part/line of SweetAlert2's codebase is causing the `Super expression must either be null or a function` issue?

**Comment by limonte** [comment]
@zenflow do you maybe had a chance to take a look on this? I was trying to figure out, no luck for now. This is quite critical to our community, even if it's not the issue on SweetAlert2's side, the vast majority of our users are using `uglify-js` for building production builds and we will be the reason of their pain trying to understand why dev environment works and production does not.

**Comment by zenflow** [comment]
Sorry for the delay.. I've investigated and found a solution (PR #1062).. I don't think we or this package is the problem, but we can do something to save the pain that uglify-js is causing with our package.

## Investigation (in case this ever comes up again)

It seems to be a problem with how two (or more) uglify-js `compress` directives (`unused` and one or more others) are interacting on our `mixin` method..

Here is the original source:

```js
function mixin (mixinParams) {
  const Swal = this
  return withNoNewKeyword(
    class MixinSwal extends Swal {
      _main (params) {
        return super._main(Object.assign({}, mixinParams, params))
      }
    }
  )
}
```

Here it is after babel transforms:

```js
function mixin(mixinParams) {
  var Swal = this;
  return withNoNewKeyword(function (_Swal) {
    inherits(MixinSwal, _Swal);

    function MixinSwal() {
      classCallCheck(this, MixinSwal);
      return possibleConstructorReturn(this, (MixinSwal.__proto__ || Object.getPrototypeOf(MixinSwal)).apply(this, arguments));
    }

    createClass(MixinSwal, [{
      key: '_main',
      value: function _main(params) {
        return get(MixinSwal.prototype.__proto__ || Object.getPrototypeOf(MixinSwal.prototype), '_main', this).call(this, _extends({}, mixinParams, params));
      }
    }]);
    return MixinSwal;
  }(Swal));
}
```

Still no problems.

Compare that to the version we get after running through `uglify-js`, with the default options but with `mangle: false, output: { beautify: true }`:

```js
function(mixinParams) {
	return withNoNewKeyword(function(_Swal) {
		function MixinSwal() {
			return classCallCheck(this, MixinSwal), possibleConstructorReturn(this, (MixinSwal.__proto__ || Object.getPrototypeOf(MixinSwal)).apply(this, arguments));
		}
		return inherits(MixinSwal, this), createClass(MixinSwal, [ {
			key: "_main",
			value: function(params) {
				return get(MixinSwal.prototype.__proto__ || Object.getPrototypeOf(MixinSwal.prototype), "_main", this).call(this, _extends({}, mixinParams, params));
			}
		} ]), MixinSwal;
	}());
}
```

The call to `inherits()` will throw because the argument in the second position, `this`, will always be `undefined`, because the containing function is called without any context.

If we disable the `unused` directive, or every other directive under `compress` *except* `unused`, we get working code. I didn't bother to figure out which of the dozens of directives is interacting with `unused` to create this problem.

Although I don't understand the problem in detail, I know that uglify-js is screwing up the reference to the parent swal (i.e. `Swal`). So I tried just changing to code to make this reference in a different way, and it worked. See PR #1062

**Comment by zenflow** [comment]
@limonte I'm not too worried, but do you have any ideas about how to protect against regressions?

We could perhaps switch back to uglify-js here, and run unit tests against the minified build rather than the non-minified, but then it would be crucial to have sourcemaps.

**Comment by limonte** [comment]
> We could perhaps switch back to uglify-js here

Agree, will do.

> but then it would be crucial to have sourcemaps.

Agree, will do in separate PR and ask for your review.

**Comment by limonte** [comment]
Fixed in [v7.24.3](https://github.com/sweetalert2/sweetalert2/releases/tag/v7.24.3)

Thanks @paolog22 @Vagabondtq for reporting and @zenflow for investigating and fixing! This was the tricky one to fix.
